lecture6/hw_ig.py

```diff
@@ -0,0 +1,67 @@
+# 1. Implement a function that flatten incoming data:
+# non-iterables and elements from iterables (any nesting depth should be supported)
+# function should return an iterator (generator function)
+# don't use third-party libraries
+# Example input:
+# a = [1, 2, 3]
+# b = 6
+# c = 'zhaba'
+# d = [[1, 2], [3, 4]]
+# Output: 1 2 3 6 z h a b a 1 2 3 4
+#
+# 2. Implement a map-like function that returns an iterator (generator function).
+# Extra functionality: if arg function can't be applied, return element as is + text exception
+# Example input:
+# a = [1, 2, 3]
+# b = 6
+# c = 'zhaba'
+# d = True
+# fun = lambda x: x[0]
+# Output:
+# 1
+# 6: 'int' object is not subscriptable
+# z
+# True: 'bool' object is not subscriptable
+
+
+def merge_elems(*elems):
+    for elem in elems:
+        if not hasattr(elem, '__iter__'):
+            yield elem
+        else:
+            if isinstance(elem, str) and len(elem) == 1:
+                yield elem
+            else:
+                for item in merge_elems(*elem):
```

Comment:
'yield from' might be used here but it's not essential :)

Reply:
Done.